pandas/core/internals/construction.py

```diff
@@ -246,10 +246,13 @@ def init_dict(data, index, columns, dtype=None):
 # ---------------------------------------------------------------------
 
 
-def prep_ndarray(values, copy=True):
+def prep_ndarray(values, copy=True) -> np.ndarray:
     if not isinstance(values, (np.ndarray, ABCSeries, Index)):
         if len(values) == 0:
             return np.empty((0, 0), dtype=object)
+        elif isinstance(values, range):
+            arr = np.arange(values.start, values.stop, values.step, dtype="int64")
```

Comment:
It seems like some care is needed here in respect to dtypes.  Specifically if the `range` contains values only supported by `uint64`, or values only supported by Python integers.

For example, the following works on `master`:
```python
In [2]: pd.DataFrame(range(2**63, 2**63 + 4))
Out[2]: 
                     0
0  9223372036854775808
1  9223372036854775809
2  9223372036854775810
3  9223372036854775811

In [3]: _.dtypes
Out[3]: 
0    uint64
dtype: object

In [4]: pd.DataFrame(range(2**73, 2**73 + 4))
Out[4]: 
                        0
0  9444732965739290427392
1  9444732965739290427393
2  9444732965739290427394
3  9444732965739290427395

In [5]: _.dtypes
Out[5]: 
0    object
dtype: object
```
But both fail with the changes in this PR:
```python
In [2]: pd.DataFrame(range(2**63, 2**63 + 4))
---------------------------------------------------------------------------
OverflowError: Python int too large to convert to C long

In [3]: pd.DataFrame(range(2**73, 2**73 + 4))
---------------------------------------------------------------------------
OverflowError: Python int too large to convert to C long
```

Admittedly, this is a bit of a corner case.  It also looks like the issue isn't limited to the PR, as the `Series` equivalent of the above fails on master.

Reply:
xref #30173 for the failing `Series` example on `master`; I'd expect that both issues could be solved in a similar way.